src/x/generics/list/list.go

```diff
@@ -0,0 +1,314 @@
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package list
+
+import (
+	"github.com/m3db/m3/src/x/pool"
+	"github.com/mauricelam/genny/generic"
+)
+
+// ValueType is the generic element type for use with the list.
+type ValueType generic.Type
+
+// Element is an element of a linked list.
+type Element struct {
+	// Next and previous pointers in the doubly-linked list of elements.
+	// To simplify the implementation, internally a list l is implemented
+	// as a ring, such that &l.root is both the next element of the last
+	// list element (l.Back()) and the previous element of the first list
+	// element (l.Front()).
+	next, prev *Element
+
+	// The list to which this element belongs.
+	list *List
+
+	// The value stored with this element.
+	Value ValueType
+}
+
+// Next returns the next list element or nil.
+func (e *Element) Next() *Element {
+	if p := e.next; e.list != nil && p != &e.list.root {
+		return p
+	}
+	return nil
+}
+
+// Prev returns the previous list element or nil.
+func (e *Element) Prev() *Element {
+	if p := e.prev; e.list != nil && p != &e.list.root {
+		return p
+	}
+	return nil
+}
+
+// List represents a doubly linked list.
+// The zero value for List is an empty list ready to use.
+type List struct {
+	root     Element // sentinel list element, only &root, root.prev, and root.next are used
+	len      int     // current list length excluding (this) sentinel element
+	ePool    *ElementPool
+	poolOpts pool.ObjectPoolOptions
+}
+
+// Init initializes or clears list l.
+func (l *List) Init() *List {
+	l.root.next = &l.root
+	l.root.prev = &l.root
+	l.len = 0
+	l.ePool = newElementPool(l.poolOpts)
+	return l
+}
+
+// New returns an initialized list.
+func New(opts pool.ObjectPoolOptions) *List {
```

Comment:
more a question: do you want to be able to pass in an element pool externally here? might make it easier to reuse pooled objects

Reply:
I don't see a scenario when someone might want to use a pool for within this list as well as some other place. Sounds like a difficult to find bug waiting to happen.

Reply:
some random cases where it could be useful to have the pool be external: the lifecycle of the pool is bigger than the list, or if the pool is shared by multiple lists, etc. essentially if you're ever in a scenario where there are multiple pools being allocated, it's probably worth considering reusing the pool.

Reply:
Fair enough, I'll change the constructor and have it create a default pool if nil is passed in.